File: pkg/twrfiles/format_air_test.go
package twrfiles

import (
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func TestFormatAIR_KBTVGolden(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("testdata", "KBTV_example.air"))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	rows, errs := ParseAIR(string(raw))
	if len(errs) != 0 {
		t.Fatalf("parse errs: %v", errs)
	}
	got := FormatAIR(rows)
	want, err := os.ReadFile(filepath.Join("testdata", "KBTV_example.formatted.air"))
	if err != nil {
		t.Fatalf("read golden: %v", err)
	}
	if got != string(want) {
		t.Errorf("FormatAIR golden mismatch\n--- got ---\n%q\n--- want ---\n%q", got, string(want))
	}
	if strings.Contains(got, "\r") {
		t.Error("formatted AIR contains CR")
	}
	if !strings.HasSuffix(got, "\n") {
		t.Error("formatted AIR missing trailing newline")
	}
	// No legend comments.
	if strings.Contains(got, ";") {
		t.Error("FormatAIR must not emit comment legend by default")
	}
}

func TestFormatAIR_RoundTripStructural(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("testdata", "KBTV_example.air"))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	rows1, errs := ParseAIR(string(raw))
	if len(errs) != 0 {
		t.Fatalf("parse1 errs: %v", errs)
	}
	text := FormatAIR(rows1)
	rows2, errs2 := ParseAIR(text)
	if len(errs2) != 0 {
		t.Fatalf("parse2 errs: %v", errs2)
	}
	if !reflect.DeepEqual(rows1, rows2) {
		t.Errorf("round-trip mismatch\n  rows1=%+v\n  rows2=%+v", rows1, rows2)
	}
	if text2 := FormatAIR(rows2); text != text2 {
		t.Error("FormatAIR not idempotent after re-parse")
	}
}

func TestFormatAIR_EmptyList(t *testing.T) {
	got := FormatAIR(nil)
	if got != "" {
		t.Errorf("nil list = %q, want empty", got)
	}
	got = FormatAIR([]Aircraft{})
	if got != "" {
		t.Errorf("empty list = %q, want empty", got)
	}
}

func TestFormatAIR_SixteenFieldsAndUppercase(t *testing.T) {
	rows := []Aircraft{
		{
			Callsign:  "aal99",
			Type:      "b738/f",
			Engine:    "j",
			Rules:     "i",
			Dep:       "kbtv",
			Arr:       "kbos",
			CruiseAlt: 29000,
			Route:     "BTV4 MPV",
			Remarks:   "/v/charts",
			Squawk:    "2200",
			XPDRMode:  "s",
			Lat:       44.469758,
			Lon:       -73.154747,
			Alt:       335,
			Speed:     0,
			Heading:   360,
		},
	}
	got := FormatAIR(rows)
	line := strings.TrimSuffix(got, "\n")
	fields := strings.Split(line, ":")
	if len(fields) != 16 {
		t.Fatalf("fields = %d, want 16: %q", len(fields), line)
	}
	if fields[0] != "AAL99" || fields[1] != "B738/F" || fields[2] != "J" || fields[3] != "I" {
		t.Errorf("upper fields wrong: %v", fields[:4])
	}
	if fields[4] != "KBTV" || fields[5] != "KBOS" {
		t.Errorf("dep/arr = %s/%s", fields[4], fields[5])
	}
	if fields[6] != "29000" {
		t.Errorf("cruise = %q", fields[6])
	}
	if fields[7] != "BTV4 MPV" || fields[8] != "/v/charts" {
		t.Errorf("route/remarks altered: %q %q", fields[7], fields[8])
	}
	if fields[10] != "S" {
		t.Errorf("xpdr = %q", fields[10])
	}
	if fields[11] != "44.469758" || fields[12] != "-73.154747" {
		t.Errorf("lat/lon = %s/%s", fields[11], fields[12])
	}
	if fields[13] != "335" || fields[14] != "0" || fields[15] != "360" {
		t.Errorf("alt/spd/hdg = %s/%s/%s", fields[13], fields[14], fields[15])
	}
}

func TestFormatAIR_CompactVsFractional(t *testing.T) {
	rows := []Aircraft{
		{
			Callsign:  "N1",
			Type:      "C172",
			Engine:    EnginePiston,
			Rules:     RulesVFR,
			Dep:       "KBTV",
			Arr:       "KLEB",
			CruiseAlt: 3500,
			Route:     "DCT",
			Remarks:   "",
			Squawk:    "1200",
			XPDRMode:  XPDRModeNormal,
			Lat:       44.5,
			Lon:       -73.1,
			Alt:       335.5, // fractional
			Speed:     90.25,
			Heading:   180.0, // whole
		},
	}
	got := FormatAIR(rows)
	line := strings.TrimSuffix(got, "\n")
	fields := strings.Split(line, ":")
	if len(fields) != 16 {
		t.Fatalf("fields = %d: %q", len(fields), line)
	}
	if fields[11] != "44.500000" || fields[12] != "-73.100000" {
		t.Errorf("lat/lon 6dp = %s %s", fields[11], fields[12])
	}
	if fields[13] != "335.5" {
		t.Errorf("alt fractional = %q", fields[13])
	}
	if fields[14] != "90.25" {
		t.Errorf("speed fractional = %q", fields[14])
	}
	if fields[15] != "180" {
		t.Errorf("heading whole = %q", fields[15])
	}
	// Route/remarks preserved empty is fine (empty field).
	if fields[8] != "" {
		t.Errorf("remarks = %q", fields[8])
	}
}

func TestFormatAIR_PreservesRouteRemarksAsStored(t *testing.T) {
	rows := []Aircraft{
		{
			Callsign:  "X1",
			Type:      "A320",
			Engine:    EngineJet,
			Rules:     RulesIFR,
			Dep:       "KBTV",
			Arr:       "KBOS",
			CruiseAlt: 10000,
			Route:     "mixed Case Route",
			Remarks:   "keep /v mixed",
			Squawk:    "1234",
			XPDRMode:  XPDRModeStandby,
			Lat:       1, Lon: 2, Alt: 3, Speed: 4, Heading: 5,
		},
	}
	got := FormatAIR(rows)
	if !strings.Contains(got, ":mixed Case Route:keep /v mixed:") {
		t.Errorf("route/remarks not preserved as-is: %q", got)
	}
}

File: webjs/airport-editor/format-air.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { parseAIR } from '../../internal/web/static/js/openfsd/airport-editor/parse-air.js';
import { formatAIR } from '../../internal/web/static/js/openfsd/airport-editor/format-air.js';
import {
  EnginePiston,
  EngineJet,
  RulesVFR,
  RulesIFR,
  XPDRModeNormal,
  XPDRModeStandby,
} from '../../internal/web/static/js/openfsd/airport-editor/model.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const fixtures = join(__dirname, '../../pkg/twrfiles/testdata');

test('formatAIR KBTV golden byte-equal', () => {
  const raw = readFileSync(join(fixtures, 'KBTV_example.air'), 'utf8');
  const { aircraft, errors } = parseAIR(raw);
  assert.equal(errors.length, 0);
  const got = formatAIR(aircraft);
  const want = readFileSync(join(fixtures, 'KBTV_example.formatted.air'), 'utf8');
  assert.equal(got, want);
  assert.ok(!got.includes('\r'));
  assert.ok(got.endsWith('\n'));
  assert.ok(!got.includes(';'));
});

test('formatAIR round-trip structural + idempotent', () => {
  const raw = readFileSync(join(fixtures, 'KBTV_example.air'), 'utf8');
  const { aircraft: r1, errors: e1 } = parseAIR(raw);
  assert.equal(e1.length, 0);
  const text = formatAIR(r1);
  const { aircraft: r2, errors: e2 } = parseAIR(text);
  assert.equal(e2.length, 0);
  assert.deepEqual(r1, r2);
  assert.equal(formatAIR(r2), text);
});

test('formatAIR empty list', () => {
  assert.equal(formatAIR(null), '');
  assert.equal(formatAIR([]), '');
});

test('formatAIR sixteen fields and uppercase', () => {
  const rows = [
    {
      callsign: 'aal99',
      type: 'b738/f',
      engine: 'j',
      rules: 'i',
      dep: 'kbtv',
      arr: 'kbos',
      cruiseAlt: 29000,
      route: 'BTV4 MPV',
      remarks: '/v/charts',
      squawk: '2200',
      xpdrMode: 's',
      lat: 44.469758,
      lon: -73.154747,
      alt: 335,
      speed: 0,
      heading: 360,
    },
  ];
  const got = formatAIR(rows);
  const line = got.replace(/\n$/, '');
  const fields = line.split(':');
  assert.equal(fields.length, 16);
  assert.equal(fields[0], 'AAL99');
  assert.equal(fields[1], 'B738/F');
  assert.equal(fields[2], 'J');
  assert.equal(fields[3], 'I');
  assert.equal(fields[4], 'KBTV');
  assert.equal(fields[5], 'KBOS');
  assert.equal(fields[6], '29000');
  assert.equal(fields[7], 'BTV4 MPV');
  assert.equal(fields[8], '/v/charts');
  assert.equal(fields[10], 'S');
  assert.equal(fields[11], '44.469758');
  assert.equal(fields[12], '-73.154747');
  assert.equal(fields[13], '335');
  assert.equal(fields[14], '0');
  assert.equal(fields[15], '360');
});

test('formatAIR compact vs fractional', () => {
  const rows = [
    {
      callsign: 'N1',
      type: 'C172',
      engine: EnginePiston,
      rules: RulesVFR,
      dep: 'KBTV',
      arr: 'KLEB',
      cruiseAlt: 3500,
      route: 'DCT',
      remarks: '',
      squawk: '1200',
      xpdrMode: XPDRModeNormal,
      lat: 44.5,
      lon: -73.1,
      alt: 335.5,
      speed: 90.25,
      heading: 180.0,
    },
  ];
  const got = formatAIR(rows);
  const fields = got.replace(/\n$/, '').split(':');
  assert.equal(fields.length, 16);
  assert.equal(fields[11], '44.500000');
  assert.equal(fields[12], '-73.100000');
  assert.equal(fields[13], '335.5');
  assert.equal(fields[14], '90.25');
  assert.equal(fields[15], '180');
  assert.equal(fields[8], '');
});

test('formatAIR preserves route/remarks as stored', () => {
  const rows = [
    {
      callsign: 'X1',
      type: 'A320',
      engine: EngineJet,
      rules: RulesIFR,
      dep: 'KBTV',
      arr: 'KBOS',
      cruiseAlt: 10000,
      route: 'mixed Case Route',
      remarks: 'keep /v mixed',
      squawk: '1234',
      xpdrMode: XPDRModeStandby,
      lat: 1,
      lon: 2,
      alt: 3,
      speed: 4,
      heading: 5,
    },
  ];
  const got = formatAIR(rows);
  assert.ok(got.includes(':mixed Case Route:keep /v mixed:'));
});

test('formatAirCompactFloat non-finite falls back', async () => {
  const { formatAirCompactFloat } = await import(
    '../../internal/web/static/js/openfsd/airport-editor/format-air.js'
  );
  assert.equal(formatAirCompactFloat(Number.NaN), '0');
  assert.equal(formatAirCompactFloat(Number.POSITIVE_INFINITY), '0');
});
